### PR TITLE
Autodiff

### PR DESCRIPTION
**Pull request by abadams**
This is the core feature from the Halide gradients paper to be presented at siggraph tomorrow. It adds the ability to take the derivatives of scalar values with respect to Funcs in a pipeline. The derivatives are a new set of synthesized Funcs that you can call, schedule, etc. The implementation uses reverse-mode automatic differentiation. See test/correctness/fit_function.cpp for an example.

This code was mostly written by @BachiLi . The commits are by me because I wanted to extract the minimum viable feature for purposes of discussion.

There's a bunch of non-idiomatic stuff to clean up still in Derivative.cpp and DerivativeUtils.cpp.

The paper is available here: https://people.csail.mit.edu/tzumao/gradient_halide/

### CODE REVIEW

**Comment by abadams** [comment]
I have done several rounds of cleanup on this now, and it could use someone else's eyes. PTAL.

File: src/Derivative.cpp [hunk 1]
```diff
@@ -0,0 +1,1568 @@
...
+    vector<string> order =
+        realization_order({ output.function() }, env).first;
+    vector<Func> funcs;
+    funcs.reserve(order.size());
+    // Internal::debug(0) << "Sorted Func list:" << "\n";
+    // for (const auto &func_name : order) {
+    //     Internal::debug(0) << "  . " << func_name << "\n";
+    // }
+    for (const auto &func_name : order) {
+        Func func(env[func_name]);
```
Comment (merrymercy) [inline src/Derivative.cpp]: useless?

File: src/Derivative.cpp [hunk 2]
```diff
@@ -0,0 +1,1568 @@
...
+        funcs.push_back(Func(env[func_name]));
```
Comment (merrymercy) [inline src/Derivative.cpp]: func is useless?

Comment (abadams) [reply src/Derivative.cpp]: Thanks. Fixed.

**Comment by BachiLi** [comment]
Some updates:
https://github.com/BachiLi/Halide/tree/autodiff
- Fix bugs in cast derivatives
- Use internal_assert everywhere instead of assert
- Override all functions in ReverseAccumulationVisitor
- A few more tests in correctness/autodiff.cpp
- Derivative for mod function
- Fix a bug in an optimization for general scattering

**Comment by BachiLi** [comment]
Fixed a few bugs in an optimization for general scattering (d_f(r.x) += ... and d_f(4 * r.x + r.y) += ...)
https://github.com/BachiLi/Halide/commit/84f54897f72a0730c1a892d5381fdc0b890b4b91
https://github.com/BachiLi/Halide/commit/a794bae362302f5f1f845fe52b3df14fe02fcf7c
https://github.com/BachiLi/Halide/commit/892ccc99a61ffde9cac79a79d2de22a53512059e

**Comment by abadams** [comment]
I sent you a contributor invite so that you can update this branch/PR directly.

**Comment by abadams** [comment]
Looks like windows is failing because there's a printf "Error: " printing an RMS error, and MSBuild interprets that to mean the build failed, for reasons. Pushing a fix.

**Comment by abadams** [comment]
All green. Needs a stamp from someone.

**Pull request merged**